File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void solve(vector<int>& nums,int index,vector<int> temp,vector<vector<int>> &ans){
        if(index >= nums.size()){
            ans.push_back(temp);
            return;
        }
        
        // include
        temp.push_back(nums[index]);
        solve(nums,index+1,temp,ans);
        
        //exclude
        temp.pop_back();
        solve(nums,index+1,temp,ans);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        
        solve(nums,0,temp,ans);
        
        return ans;
        
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {

    public void solve(int[] nums, int start, List<Integer> temp, List<List<Integer>> ans){
        ans.add(new ArrayList<>(temp));

        for(int i = start; i < nums.length; i++){
            temp.add(nums[i]);
            solve(nums, i + 1, temp, ans);
            temp.remove(temp.size() - 1);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
      List<List<Integer>> ans = new ArrayList<>();
      solve(nums, 0, new ArrayList<>(), ans);
      return ans;  
    }
}